include/ignition/math/SpeedLimiter.hh

```diff
@@ -0,0 +1,147 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/*
+ * Author: Enrique Fernández
+ * Modified: Carlos Agüero
+ */
+
+#ifndef IGNITION_MATH_SYSTEMS_SPEEDLIMITER_HH_
+#define IGNITION_MATH_SYSTEMS_SPEEDLIMITER_HH_
+
+#include <chrono>
+#include <memory>
+#include <ignition/math/config.hh>
+#include "ignition/math/Helpers.hh"
+
+namespace ignition
+{
+namespace math
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_MATH_VERSION_NAMESPACE {
+  // Forward declaration.
+  class SpeedLimiterPrivate;
+
+  /// \brief Class to limit velocity, acceleration and jerk.
+  /// \ref https://github.com/ros-controls/ros_controllers/tree/melodic-devel/diff_drive_controller
+  class IGNITION_MATH_VISIBLE SpeedLimiter
+  {
+    /// \brief Constructor.
+    /// \param [in] _hasVelocityLimits     If true, applies velocity limits.
+    /// \param [in] _hasAccelerationLimits If true, applies acceleration limits.
+    /// \param [in] _hasJerkLimits         If true, applies jerk limits.
+    /// \param [in] _minVelocity Minimum velocity [m/s], usually <= 0. Defaults
+    ///     to negative infinity.
+    /// \param [in] _maxVelocity Maximum velocity [m/s], usually >= 0. Defaults
+    ///     to positive infinity.
+    /// \param [in] _minAcceleration Minimum acceleration [m/s^2], usually <= 0.
+    ///     Defaults to negative infinity.
+    /// \param [in] _maxAcceleration Maximum acceleration [m/s^2], usually >= 0.
+    ///     Defaults to positive infinity.
+    /// \param [in] _minJerk Minimum jerk [m/s^3], usually <= 0. Defaults to
+    ///     negative infinity.
+    /// \param [in] _maxJerk Maximum jerk [m/s^3], usually >= 0. Defaults to
+    ///     positive infinity.
+    public: SpeedLimiter(
+        bool   _hasVelocityLimits     = false,
+        bool   _hasAccelerationLimits = false,
+        bool   _hasJerkLimits         = false,
+        double _minVelocity           = -INF_D,
+        double _maxVelocity           = INF_D,
+        double _minAcceleration       = -INF_D,
+        double _maxAcceleration       = INF_D,
+        double _minJerk               = -INF_D,
+        double _maxJerk               = INF_D);
```

Comment:
I'm not a huge fan of APIs that use big long lists of values like this because I usually forget and have to look up the order in which they should be used.

I wonder if we need the `bool` values, since the min and max values now default to unlimited? We could create private `bool` values in the constructor by inspecting the min and max values to see if any are finite.

Aside from that, I don't have any better suggestions for the API